### DISCUSSION TITLE
Migrating to ParseSwiftOG: From Parse-SDK-iOS-OSX

### DISCUSSION THREAD

**Discussion by cbaker6** [netreconlab/Parse-Swift#71] [General]
`Parse-Swift` provides a modernized Swift SDK that allows developers to take advantage of the latest and greatest Swift and SwiftUI has to offer (SwiftUI, Combine, async/await, etc). In addition, Parse-Server 6.0.0+ has been released which has a number of feature additions for querying, live-queries, etc. Parse-Swift<sup>OG</sup> is miles ahead of the `Parse-SDK-iOS-OSX` and by my estimates (I have a solid understanding of all of the major Parse Client SDK's), **I would say Parse-Swift<sup>OG</sup> 2+ years ahead of `Parse-SDK-iOS-OSX` in development time.** There's one feature the `Parse-SDK-iOS-OSX` SDK has that `Parse-Swift` doesn't and that's a local CoreData store. The CoreData store in `Parse-SDK-iOS-OSX` has a number of noted issues (more in #69) and can't be used when cache is enabled. The caching system in `Parse-Swift` is native and far better than the custom caching system in `Parse-SDK-iOS-OSX`. I discuss a better way to Parse-Swift with local storage in #69.

Migrating your project from `Parse-SDK-iOS-OSX` to `Parse-Swift` will take some work as the frameworks are completely different along with the design philosophy (see more in [README](https://github.com/netreconlab/Parse-Swift/blob/main/README.md)). Nonetheless, migrating your project is possible (see [discussion](https://community.parseplatform.org/t/how-to-transition-from-the-old-parse-framework/2807)). After converting the code, you will most likely want to access the Keychain from `Parse-SDK-iOS-OSX` to provide a seamless experience for your users. 

`Parse-Swift` makes this easy, please see the information [here](https://community.parseplatform.org/t/how-to-transition-from-the-old-parse-framework/2807/21?u=cbaker6) and [here](https://github.com/parse-community/Parse-Swift/pull/391) to learn about the methods available to migrate to `Parse-Swift`. An example application that demonstrates how to convert from the Objective-C SDK to the Swift SDK can be found at [ParseMigrateKeychain](https://github.com/netreconlab/ParseMigrateKeychain). Specifically, look at [ContentViewModel.swift](https://github.com/netreconlab/ParseMigrateKeychain/blob/main/ParseMigrateKeychain/ContentViewModel.swift) for code examples. The basic steps are below:

1. Open your project, click on your project in the `Project Navigator` panel
1. Select the name of your project under Project in the right panel
1. Click the `Package Dependencies` tab
1. Add the https://github.com/netreconlab/Parse-Swift/ and for `Dependency Rule`, select `Up to Next Major Version` and input `5.1.1` for the specific version and `< 6.0.0`
1. If the Objective-C SDK was already added via SPM, skip the next 2 steps
1. Remove the Objective-C SDK from your project if it was added via Cocoapods
1. Add the Objective-C SDK to your project using SPM. You can use this [repo](https://github.com/mman/Parse-SDK-iOS-OSX/tree/spm) as you will only need `PFUser` and `PFInstallation`
1. Convert your app from using the Objective-C SDK to the Swift SDK
1. Create a `User` and `Installation` [model](https://github.com/netreconlab/ParseMigrateKeychain/tree/main/ParseMigrateKeychain/Models) that conform to `ParseObject`
1. Migrate `PFUser->User` using [loginUsingObjCKeychain](https://swiftpackageindex.com/parse-community/parse-swift/4.14.2/documentation/parseswift/parseuser/loginusingobjckeychain(options:))
1. Migrate `PFInstallation->Installation` using [become](https://swiftpackageindex.com/parse-community/parse-swift/4.14.2/documentation/parseswift/parseinstallation/become(_:copyentireinstallation:options:))
1. Remove the Objective-C SDK from your project after you have migrated all users

**Comment by cbaker6** [comment]
@ashish-naik if you are migrating your project from the Objective-C SDK, this discussion may be of interest to you.

**Reply by ashish-naik** [reply]
I am interested but not immediately.  I have added to my "want to use" list :)

**Reply by cbaker6** [reply]
Not a problem, be sure to also checkout #72 to see the feature comparisons with the other Parse SDK's
